core/src/main/scala/akka/kafka/internal/TransactionalSource.scala

```diff
@@ -138,11 +141,12 @@ private[kafka] final class TransactionalSource[K, V](consumerSettings: ConsumerS
         inFlightRecords.add(Map(new TopicPartition(rec.topic(), rec.partition()) -> rec.offset()))
 
       override protected def stopConsumerActor(): Unit =
-        sourceActor.ref.tell(Drain(Some(consumerActor), KafkaConsumerActor.Internal.Stop), sourceActor.ref)
+        sourceActor.ref.tell(Drain(inFlightRecords.assigned(), Some(consumerActor), KafkaConsumerActor.Internal.Stop),
+                             sourceActor.ref)
 
       // This is invoked in the KafkaConsumerActor thread when doing poll.
       override def partitionRevokedHandler(revokedTps: Set[TopicPartition]): Unit = {
-        if (waitForDraining()) {
+        if (waitForDraining(revokedTps)) {
```

Comment:
AFAIK we always get all partitions revoked, and then get assigned again the ones, that are kept to the same consumer. But anyways, this can come in anticipating future Kafka client improvements.